samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,234 @@
+	if (retval != 0) {
+		printk("download_client_start() failed, err %d",
+			retval);
+		return;
+	}
+}
+
+
+static int app_download_client_event_handler(
+	struct download_client *const dfu,
+	enum download_client_evt event, u32_t error)
+{
+	int err;
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		struct flash_pages_info info;
+		static u32_t address = CONFIG_FLASH_WRITE_OFFSET;
+
+		err = flash_get_page_info_by_offs(flash_dev, address, &info);
+		if (err != 0) {
+			printk("flash_get_page_info_by_offs returned error %d\n",
+				err);
+			return 1;
+		}
+		flash_write_protection_set(flash_dev, false);
+		if (!is_flash_page_erased[info.index]) {
+			err = flash_erase(flash_dev, address, info.size);
+			if (err != 0) {
+				printk("flash_erase returned error %d at address %08x\n",
+					err, address);
+				return 1;
+			}
+			is_flash_page_erased[info.index] = true;
+		}
+		err = flash_write(flash_dev, address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			printk("Flash write error %d at address %08x\n",
+				err, address);
+			return 1;
+		}
+		address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
+		printk("DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE");
+		break;
+
+	case DOWNLOAD_CLIENT_EVT_ERROR: {
+		download_client_disconnect(dfu);
+		printk("Download client error, please restart "
+				"the application\n");
+		break;
+	}
+	default:
+		break;
+	}
+	return 0;
+}
+
+
+/**@brief Turn on LED0 and LED1 if CONFIG_APPLICATION_VERSION
+ * is 2 and LED0 otherwise.
+ */
+static void led_app_version(void)
+{
+	struct device *dev;
+
+	dev = device_get_binding(LED_PORT);
+	if (dev == 0) {
+		printk("Nordic nRF GPIO driver was not found!\n");
+		return;
+	}
+
+	gpio_pin_configure(dev, LED0_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED0_GPIO_PIN, 1);
+
+#if CONFIG_APPLICATION_VERSION == 2
+	gpio_pin_configure(dev, LED1_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED1_GPIO_PIN, 1);
+#endif
+}
+
+void dfu_button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	start_dfu = true;
+}
+
+static void dfu_button_init(void)
+{
+	int err;
+	struct device *gpiob;
+
+	gpiob = device_get_binding(SW0_GPIO_CONTROLLER);
+	if (gpiob == 0) {
+		printk("Nordic nRF GPIO driver was not found!\n");
+		return;
+	}
+	err = gpio_pin_configure(gpiob, SW0_GPIO_PIN,
+			GPIO_DIR_IN		|
+			GPIO_INT		|
+			GPIO_PUD_PULL_UP|
+			GPIO_INT_EDGE	|
+			GPIO_INT_ACTIVE_LOW);
+	if (err == 0) {
+		gpio_init_callback(&gpio_cb, dfu_button_pressed,
+			BIT(SW0_GPIO_PIN));
+		err = gpio_add_callback(gpiob, &gpio_cb);
+	}
+	if (err == 0) {
+		err = gpio_pin_enable_callback(gpiob, SW0_GPIO_PIN);
+	}
+	if (err != 0) {
+		printk("Unable to configure SW0_GPIO_PIN!\n");
+	}
+}
+
+static void application_update(void)
+{
+	start_dfu = false;
+	app_dfu_init();
+	app_dfu_transfer_start();
+
+	while (dfu.status != DOWNLOAD_CLIENT_STATUS_DOWNLOAD_COMPLETE) {
```

Comment:
What if an error occurs? You are calling `download_client_disconnect` from the event handler and the status will be `DOWNLOAD_CLIENT_STATUS_IDLE`. You will never exit this loop.

Reply:
I am only calling `download_client_disconnect `from the event handler when there is an error but I will change the while above to `while(true)` and have a switch inside the while.

Reply:
I have now rewritten the while and also check the two possible errors
`DOWNLOAD_CLIENT_STATUS_HALTED`
`DOWNLOAD_CLIENT_STATUS_DOWNLOAD_COMPLETE`